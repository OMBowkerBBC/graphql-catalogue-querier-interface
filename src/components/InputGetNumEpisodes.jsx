import { useState } from 'react'
import Input from '@bbc/igm-input'
import '@bbc/igm-input/dist/Input.css'
import PropTypes from 'prop-types'

const InputGetNumEpisodes = ({ setQueryArguments }) => {
  const [showErrorMessage, setShowErrorMessage] = useState(false)
  const [inputValue, setInputValue] = useState('')

  const validateInput = (userInput) => {
    setInputValue(userInput)
    if (isNaN(parseInt(userInput)) && userInput.length > 0) {
      setShowErrorMessage(true)
    } else {
      setQueryArguments({amount: parseInt(userInput)})
      setShowErrorMessage(false)
    }
  }

  return (
    <Input
      name='amount'
      placeholder='Amount...'
      isClearable={true}
      label='Number of episodes to retrieve.'
      onChange={(e) => validateInput(e)}
      errorTitle={showErrorMessage ? 'Error!' : ''}
      errorMessage='Must be a number.'
      data-testid='input'
      value={inputValue}
    />
  )
}

InputGetNumEpisodes.propTypes = {
  setQueryArguments: PropTypes.func.isRequired,
}

export default InputGetNumEpisodes
