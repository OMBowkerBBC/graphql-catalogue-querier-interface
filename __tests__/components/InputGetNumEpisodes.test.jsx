import { fireEvent, render, screen } from '@testing-library/react'
import renderer from 'react-test-renderer'
import InputGetNumEpisodes from '../../src/components/InputGetNumEpisodes'

describe('GetNumEpisodes test suite', () => {
  let mockedSetQueryArguments

  describe('snapshot tests', () => {
    it('should match starting snapshot', () => {
      const tree = renderer.create(<InputGetNumEpisodes setQueryArguments={vi.fn()} />).toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  beforeEach(() => {
    mockedSetQueryArguments = vi.fn()
    render(<InputGetNumEpisodes setQueryArguments={mockedSetQueryArguments}/>)
  })
  it('should render with correct label, placeholder text and no error message', () => {
    expect(screen.getByText('Number of episodes to retrieve.')).toBeDefined()
    expect(screen.getByPlaceholderText('Amount...')).toBeDefined()
    expect(() => screen.getByText('Error')).toThrow()
  })

  it('should call setQueryArguments with valid input', () => {
    const input = screen.getByTestId('input')
    fireEvent.change(input, { target: { value: '10' } })
    expect(mockedSetQueryArguments).toHaveBeenCalled()
    expect(() => screen.getByText(/Error/)).toThrow()
  })

  it('should show error message with invalid input', () => {
    const input = screen.getByTestId('input')
    fireEvent.change(input, { target: { value: 'abc' } })
    expect(mockedSetQueryArguments).not.toHaveBeenCalled()
    expect(screen.getByText(/Error/)).toBeDefined()
  })

  it('should remove error message when invalid input is removed', () => {
    const input = screen.getByTestId('input')
    fireEvent.change(input, { target:{ value: 'abc' } })
    expect(screen.getByText(/Error/)).toBeDefined()
    fireEvent.change(input, { target:{ value: '' } })
    expect(() => screen.getByText(/Error/)).toThrow()
  })
})
