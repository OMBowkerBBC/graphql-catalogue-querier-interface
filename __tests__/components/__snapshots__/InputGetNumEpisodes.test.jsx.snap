// Vitest Snapshot v1

exports[`GetNumEpisodes test suite > snapshot tests > should match starting snapshot 1`] = `
<div
  className="css-n222kl"
>
  <div
    className="css-1vjiq4b"
  >
    <label
      className="css-xl6oz7"
      htmlFor="input-c50fsjdr4d"
      id="label-c50fsjdr4d"
    >
      Number of episodes to retrieve.
    </label>
  </div>
  <div
    className="css-130g1t6"
    disabled={false}
  >
    <input
      aria-invalid={false}
      className="css-1et5r9b"
      data-testid="input"
      disabled={false}
      id="input-c50fsjdr4d"
      name="amount"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Amount..."
      required={true}
      type="text"
      value=""
    />
  </div>
  <div
    aria-live="polite"
  />
  <div
    aria-live="polite"
  />
</div>
`;
